README.md
# cl-aima

*AIMA (AI: A Modern Approach) Code for Common Lisp*

The code here is the as that which appears in the following locations:
 * http://www.cs.berkeley.edu/~russell/code/code.tar.gz
 * http://aima.cs.berkeley.edu/lisp/code.tar.gz

More information about this code is available here:

 * http://aima.cs.berkeley.edu/lisp/doc/overview.html

The changes made are discussed in the file ``ChangeLog`` in this directory.


## Setup

Download the code and then start up SBCL:

```bash
$ git clone git@github.com:oubiwann/cl-aima.git
$ cd cl-aima
$ sbcl
```

Load the code, compile it, and run the tests:

```cl
* (load "aima.lisp")
...
T
* (aima-load 'all)
...
T
* (aima-compile)
...
T
* (test 'all)
...
 0 errors on system LANGUAGE
0
```


## Usage

At this point, you're ready to use the code that has been defined in the
various ``.lisp`` files. Looking at the tests may be helpful, for instance,
take a look at the test file ``agents/test-agents.lisp``.

In it, you will see an example for having an agent run in the "Wumpus world"
environment. You can call that directly:

```cl
* (run-environment (make-wumpus-world :max-steps 30))
```

For this exanple, you will see a series of Wumpus screens like so:

```
...
At Time step 6:
Agent [1 = -10006] perceives (NIL BREEZE NIL NIL NIL)
      and does (TURN LEFT)
    |----|----|----|----|----|----|
 5  | #  | #  | #  | #  | #  | #  |
    |----|----|----|----|----|----|
 4  | #  | O  | O  | W> | O  | #  |
    |----|----|----|----|----|----|
 3  | #  | O  |    |    |    | #  |
    |----|----|----|----|----|----|
 2  | #  | 1O | O  |    |    | #  |
    |----|----|----|----|----|----|
 1  | #  |    |    | O  | $  | #  |
    |----|----|----|----|----|----|
 0  | #  | #  | #  | #  | #  | #  |
    |----|----|----|----|----|----|
      0    1    2    3    4    5
#<WUMPUS-WORLD; Step: 6, Agents: [1 = -10006]>
...

```


ChangeLog
Wed Dec 30 18:10:21 CST 2015 Duncan M. McGreggor <oubiwann@gmail.com>

Updated the code to work with SBCL. This involved simple changes such as
converting DEFCONSTANT usage to DEFPARAMETER and commenting out a DECLARE usage
for +no-bindings+.


Thu Oct 06 15:03:34 2005  Charles A. Cox  <cox@DDD>

  General description:

    (1) shadow print-structure symbol which conflicts with Allegro Common
    Graphics.

    (2) AIMA code was disabling package locks.  It's better for the user
    to keep this mechanism activated.  It turns out that AIMA only needs
    to add one new definition to a locked symbol, so the package lock is
    now only turned off just for that one symbol.

    (3) Check if Lisp is case-sensitive.  Some AIMA code is
    case-insensitive.  While I'm sure it would be fairly easy to make AIMA
    run in a case-sensitive Lisp, it would involve many more changes than I
    felt I should make at this time.

  Specifics:

    * aima.lisp: Shadow symbol print-structure which conflicts with an
    existing symbol in Allegro Common Graphics.  Also, signal user if
    Lisp is case-sensitive since some AIMA code depends on
    case-insensitivity.

    * search/algorithms/iterative.lisp: Remove undefined function
    warning by adding an Allegro declaration.

    * utilities/cltl2.lisp: Remove Allegro package-lock override.

    * utilities/utilities.lisp: Remove Allegro package-lock override.
    Define 'while' inside Allegro without-package-locks.  Also, use
    special-operator-p instead of obsolete special-form-p.

0.99 AIMA Code, Appomattox Version, 09-Apr-2002
    Updated Franz-specific conditional code to work with newest version.
0.98 AIMA Code, Thelonious Monk Birthday Version, 10-Oct-2001
    Line 190 of agents/environments/grid-env.lisp changed to
    ((xy-p where) (parse-whats env where whats))
    Previously it had the parameter order confused. Thanks to Yoshihiro Ota.
0.97 AIMA Code, Shot Heard 'Round the World Day Version, 03-Oct-2001
    Removed support of Lispworks from the documentation. Lispworks appears to have a bug with the treatment of special variables (it shows up in logic/algorithims/tell-ask.lisp). If someone sends me a workaround I'll add it. For now, I'll just limit official support to Allegro and MCL (although you certainly can get the code to work in other Lisps).

0.96 AIMA Code, Palindrome Day Version, 10-02-2001
    Replaced get-setf-method with get-setf-expansion, as per the ANSI standard. Updated documentation. The biggest error was that I didn't mention that you need to do a aima-load before the aima-compile. Tested code with Allegro, Lispworks, and MCL. It loads, compiles, and tests out ok in all three Lisps.

0.95 AIMA Code, Patriots Day Version, 21-Apr-1997
    Four main changes: First, modified much of the code to use an object-oriented approach using defmethod. We limited this to single-inheritance, and single dispatch (on the first argument). This is both because we wanted it to work even in old Lisps that don't have a CLOS (Common Lisp Object System) implementation, and because we wanted it to be familiar to programmers who are used to object-oriented approaches in other languages such as Java or Smalltalk. Second, added extensive machine-generated HTML documentation, of which this mpage is part. Third, the directory structure was re-organized. Fourth, completed the code for a chart parser and grammar in the language directory. There were also some minor bug fixes, e.g. David Palmer pointed out that the wumpus agent can pick up the wumpus when it is trying to pick up the gold.

0.94 AIMA Code, Ground Hog Day Version, 2-Feb-1995
    Ray Liere and David Palmer pointed out that non-square grid-environments were incorrectly initialized in grid-env.lisp. Roger Kirchner pointed out that the precedence of the | operator in infix.lisp was wrong. John Sterling discovered some files where we still referenced the USER package, and suggested removing the (declare (ignore contained?)) from DEFSETF GRID-CONTENTS to avoid a warning in MCL. The infix reader was enhanced to turn lowercase symbols into variables, i.e. "F(x,C)" turns into (F ?X C). Some code that answers exercises from the book was moved out of the code repository and into the soon-to-be-released instructor's manual. Added file "index.txt", which lists all functions, types, and variables, the page they appear on in the book, and the file they appear in.

0.93 AIMA Code, Oral Roberts' Birthday Version, 24-Jan-1995
    There were problems with older Common Lisps stemming from the difference between the USER and CL-USER packages, so I just decided to eliminate all code that deals with packages. There were a few other changes to fix problems in Clisp. When compiling under Clisp, if you get an error message that says a macro is being redefined, just type "continue". After you've compiled once, you won't get the message again when you load. The code has now been tested in Harlequin, Lucid, Franz, AKCL and Clisp.

0.92 AIMA Code, Martin Luther King Day Version, 16-Jan-1995
    Added new game-playing and logic code, and new systems for learning and uncertainty. Added some documentation and minor changes to agents and search code.

0.91 AIMA Code, Newer Years Version, 5-Jan-1995
    Fixed a few minor bugs that resulted from too-quickly merging two versions of the code, and added some (declare (ignore variable))s. Tested in several implementations of Common Lisp: Harlequin, Lucid, Franz, and AKCL.

0.9 AIMA Code, New Years Version, 4-Jan-1995
    A preliminary version, released because several instructors were starting courses this week. Only the AGENTS and SEARCH systems are supported, although unsupported code for other systems is included.


aima.lisp
;;;  -*- Mode: Lisp; Syntax: Common-Lisp -*- File: aima.lisp

;;;; Vendor-Specific Customizations

#+Lucid (setq *warn-if-no-in-package* nil)

;; 6oct05 charley cox
;; print-structure symbol conflicts with existing symbol in Allegro Common
;; Graphics
#+Allegro (shadow 'print-structure)

;;;; A minimal facility for defining systems of files

(defparameter *aima-root* (truename ".") ; <<<<<<<< Edit this <<<<<<
  "The root directory where the code is stored.")

(defparameter *aima-binary-type*
  (pathname-type (compile-file-pathname "foo.lisp"))
  "If calling aima-load loads your source files and not your compiled
  binary files, insert the file type for your binaries before the <<<<
  and load systems with (aima-load-binary NAME).")

(defparameter *aima-version*
  "0.99 AIMA Code, Appomattox Version, 09-Apr-2002")

(defparameter *aima-system-names* nil
  "A list of names of the systems that have been defined.")

(defstruct aima-system
  name (requires nil) (doc "") (parts nil) (examples nil) (loaded? nil))

;;;; The Top-Level Functions:

(defmacro def-aima-system (name requires doc &body parts)
  "Define a system as a list of parts.  A part can be a string, which denotes
  a file name; or a symbol, which denotes a (sub)system name; or a list of the
  form (subdirector y/ part...), which means the parts are in a subdirectory.
  The REQUIRES argument is a list of systems that must be loaded before this
  one.  Note that a documentation string is mandatory."
  `(add-aima-system :name ',name
            :requires ',requires :doc ',doc :parts ',parts))

(defun aima-load (&optional (name 'all))
  "Load file(s), trying the system-dependent method first."
  (operate-on-aima-system name 'load-something))

(defun aima-load-binary (&optional (name 'all))
  "Load file(s), prefering binaries to source."
  (operate-on-aima-system name 'load-binary))

(defun aima-compile (&optional (name 'everything))
  "Compile (and load) the file or files that make up an AIMA system."
  (operate-on-aima-system name 'compile-load))

(defun aima-load-if-unloaded (name)
  (let ((system (get-aima-system name)))
    (unless (and system (aima-system-loaded? system))
      (aima-load system))
    system))

;;;; Support Functions

(defun add-aima-system (&key name requires doc parts examples)
  (pushnew name *aima-system-names*)
  (setf (get 'aima-system name)
    (make-aima-system :name name :examples examples
              :requires requires :doc doc :parts parts)))

(defun get-aima-system (name)
  "Return the system with this name.  (If argument is a system, return it.)"
  (cond ((aima-system-p name) name)
    ((symbolp name) (get 'aima-system name))
    (t nil)))

(defun operate-on-aima-system (part operation &key (path nil) (load t)
                    (directory-operation #'identity))
  "Perform the operation on the part (or system) and its subparts (if any).
  Reasonable operations are load, load-binary, compile-load, and echo.
  If LOAD is true, then load any required systems that are unloaded."
  (let (system)
    (cond
     ((stringp part) (funcall operation (aima-file part :path path)))
     ((and (consp part) (eq (second part) '/))
      (let* ((subdirectory (mklist (first part)))
         (new-path (append path subdirectory)))
    (funcall directory-operation new-path)
    (dolist (subpart (nthcdr 2 part))
      (operate-on-aima-system subpart operation :load load
                  :path new-path
                  :directory-operation directory-operation))))
     ((consp part)
      (dolist (subpart part)
    (operate-on-aima-system subpart operation :load load :path path
                :directory-operation directory-operation)))
     ((setf system (get-aima-system part))
      ;; Load the required systems, then operate on the parts
      (when load (mapc #'aima-load-if-unloaded (aima-system-requires system)))
      (operate-on-aima-system (aima-system-parts system) operation
                  :load load :path path
                  :directory-operation directory-operation)
      (setf (aima-system-loaded? system) t))
     (t (warn "Unrecognized part: ~S in path ~A" part path)))))

(defun aima-file (name &key (type nil) (path nil))
  "Given a file name and maybe a file type and a relative path from the
  AIMA directory, return the right complete pathname."
  (make-pathname :name name :type type :defaults *aima-root*
         :directory (append (pathname-directory *aima-root*)
                    (mklist path))))

#-MCL ;; Macintosh Common Lisp already defines this function
(defun compile-load (file)
  "Compile file and then load it."
  ;; This could be made more sophisticated, to compile only when out of date.
  (compile-file (file-with-type file "lisp"))
  (load-binary file))

(defun load-binary (file)
  "Load file, trying the binary first, but loading the source if necessary."
  (load-something file '(binary nil "lisp")))

(defun load-something (file &optional (types '(nil binary "lisp")))
  "Try each of the types in turn until we get a file that loads.
  Complain if we can't find anything.  By default, try the system-dependent
  method first, then the binary, and finally the source (lisp) file."
  (dolist (type types (warn "Can't find file: ~A" file))
    (when (load (file-with-type file type) :if-does-not-exist nil)
      (return t))))

(defun file-with-type (file type)
  "Return a pathname with the given type."
  (if (null type)
      file
    (merge-pathnames
     (make-pathname :type (if (eq type 'binary) *aima-binary-type* type))
     file)))

(defun mklist (x)
  "If x is a list, return it; otherwise return a singleton list, (x)."
  (if (listp x) x (list x)))

;;; ----------------------------------------------------------------------
;;;; Definitions of Systems
;;; ----------------------------------------------------------------------

(def-aima-system utilities ()
  "Basic functions that are loaded every time, and used by many other systems."
  ("utilities" / "utilities" "binary-tree" "queue" "cltl2" "test-utilities"))

(def-aima-system agents (utilities)
  "Code from Part I: Agents and Environments"
  ("agents" / "test-agents"
   ("environments" / "basic-env" "grid-env" "vacuum" "wumpus")
   ("agents" / "agent" "vacuum" "wumpus")
   ("algorithms" / "grid")))

(def-aima-system search (agents)
  "Code from Part II: Problem Solving and Search"
  ("search" / "test-search"
   ("algorithms" / "problems" "simple" "repeated"
    "csp" "ida" "iterative" "sma" "minimax")
   ("environments" / "games" "prob-solve")
   ("domains" / "cannibals" "ttt" "cognac" "nqueens" "path-planning"
    "puzzle8" "route-finding" "tsp" "vacuum")
   ("agents" / "ps-agents" "ttt-agent")))

(def-aima-system logic (agents)
  "Code from Part III: Logic, Inference, and Knowledge Representation"
   ("logic" / "test-logic"
    ("algorithms" / "tell-ask" "unify" "normal" "prop" "horn" "fol" "infix")
    ("environments" / "shopping")))

(def-aima-system planning ()
  "Code from Part IV: Planning and Acting"
   ("planning" / ))

(def-aima-system uncertainty (agents)
  "Code from Part V: Uncertain Knowledge and Reasoning"
  ("uncertainty" / "test-uncertainty"
   ("agents" / "mdp-agent")
   ("domains" / "mdp" "4x3-mdp")
   ("environments" / "mdp")
   ("algorithms" / "dp" "stats")))

(def-aima-system learning (uncertainty)
  "Code from Part VI: Learning"
   ("learning" / "test-learning"
    ("algorithms" / "inductive-learning" "learning-curves" "dtl" "dll"
     "nn" "perceptron" "multilayer" "q-iteration")
    ("domains" / "restaurant-multivalued" "restaurant-real"
     "restaurant-boolean" "majority-boolean" "ex-19-4-boolean"
     "and-boolean" "xor-boolean" "4x3-passive-mdp")
    ("agents" / "passive-lms-learner" "passive-adp-learner"
     "passive-td-learner" "active-adp-learner" "active-qi-learner"
     "exploring-adp-learner" "exploring-tdq-learner")))

(def-aima-system language (logic)
  "Code from Part VII, Chapters 22-23: Natural Language and Communication"
   ("language" / "test-language"
    ("algorithms" / "chart-parse")
    ("domains" / "grammars" )))

(def-aima-system all ()
  "All systems except the utilities system, which is always already loaded"
  agents search logic planning uncertainty learning language)

(def-aima-system everything ()
  "All the code, including the utilities"
  utilities all)

(setf *aima-system-names* (nreverse *aima-system-names*))

;;;; Always load the utilities

(when (string= "nIL" (symbol-name (read-from-string "nIL")))
  (cerror
   "Continue loading AIMA code."
   "Lisp reader is case-sensitive.  Some AIMA code may not work correctly."))

(aima-load 'utilities)
